index.js

```diff
@@ -91,29 +91,61 @@ function init (ssb, config) {
 
   /* start listeners */
   const rebuildManager = new RebuildManager(ssb)
-  listen.addMember(ssb, m => {
-    const { root, groupKey } = m.value.content
-    ssb.get({ id: root, meta: true }, (err, groupInitMsg) => {
-      if (err) throw err
+  const processedNewAuthors = {}
+  pull(
+    listen.addMember(ssb),
+    pull.asyncMap((m, cb) => {
+      const { root, groupKey } = m.value.content
+      ssb.get({ id: root, meta: true }, (err, groupInitMsg) => {
+        if (err) throw err
 
-      const groupId = buildGroupId({ groupInitMsg, groupKey })
-      const authors = unique([
-        groupInitMsg.value.author,
-        m.value.author,
-        ...m.value.content.recps.filter(isFeed)
-      ])
+        const groupId = buildGroupId({ groupInitMsg, groupKey })
+        const authors = unique([
+          groupInitMsg.value.author,
+          m.value.author,
+          ...m.value.content.recps.filter(isFeed)
+        ])
+
+        const record = keystore.group.get(groupId)
+        // if we haven't been in the group since before, register the group
+        if (record == null) {
+          return keystore.group.register(groupId, { key: groupKey, root }, (err) => {
+            if (err) return cb(err)
+            processAuthors(cb)
+          })
+        } else {
+          processAuthors(cb)
+        }
 
-      keystore.processAddMember({ groupId, groupKey, root, authors }, (err, newAuthors) => {
-        if (err) throw err
-        if (newAuthors.length) {
-          state.newAuthorListeners.forEach(fn => fn({ groupId, newAuthors }))
+        function processAuthors (cb) {
+          if (processedNewAuthors[groupId] === undefined) processedNewAuthors[groupId] = []
```

Comment:
done in 658aebd now. but idk it feels a tad more confusing now since sets and arrays are mixed. deduping will probably be relevant at some point though